lib/types.nix

```
@@ -243,7 +243,7 @@ rec {
     path = mkOptionType {
       name = "path";
       # Hacky: there is no ‘isPath’ primop.
-      check = x: builtins.substring 0 1 (toString x) == "/";
+      check = x: builtins.substring 0 1 (toString (x.origSrc or x)) == "/";
```

Perhaps something like this, for a more general and decoupled solution. Needs a new attr in cleanSource and friends.

```suggestion
      check = x: x._isPath or builtins.substring 0 1 (toString x) == "/";
```